src/Updater/PuliStrategy.php

```diff
@@ -0,0 +1,130 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Puli\Cli\Updater;
+
+use Humbug\SelfUpdate\Exception\HttpRequestException;
+use Humbug\SelfUpdate\Strategy\StrategyInterface;
+use Humbug\SelfUpdate\Updater;
+use Humbug\SelfUpdate\VersionParser;
+use Puli\Cli\PuliApplicationConfig;
+
+class PuliStrategy implements StrategyInterface
+{
+    const ANY = 'any';
+    const STABLE = 'stable';
+    const UNSTABLE = 'unstable';
+
+    const MANIFEST = 'https://puli.io/download/versions.json';
+    const REMOTE_PHAR = 'https://puli.io/download/%s/puli.phar';
+
+    /**
+     * @var string
+     */
+    private $stability = self::STABLE;
+
+    /**
+     * Download the remote Phar file.
+     *
+     * @param Updater $updater
+     */
+    public function download(Updater $updater)
+    {
+        /* Switch remote request errors to HttpRequestExceptions */
+        set_error_handler(array($updater, 'throwHttpRequestException'));
+
+        $remoteUrl = sprintf(
+            self::REMOTE_PHAR,
+            $this->getCurrentRemoteVersion($updater)
+        );
+
+        $result = humbug_get_contents($remoteUrl);
+        restore_error_handler();
+
+        if (false === $result) {
+            throw new HttpRequestException(sprintf(
+                'Request to URL failed: %s', $remoteUrl
+            ));
+        }
+
+        file_put_contents($updater->getTempPharFile(), $result);
+    }
+
+    /**
+     * Retrieve the current version available remotely.
+     *
+     * @param Updater $updater
+     * 
+     * @return string
+     */
+    public function getCurrentRemoteVersion(Updater $updater)
+    {
+        /* Switch remote request errors to HttpRequestExceptions */
+        set_error_handler(array($updater, 'throwHttpRequestException'));
+        $versions = json_decode(humbug_get_contents(self::MANIFEST), true);
+        restore_error_handler();
+
+        if (false === $versions) {
+            throw new HttpRequestException(sprintf(
+                'Request to URL failed: %s', self::MANIFEST
+            ));
+        }
+
+        $versionParser = new VersionParser($versions);
+
+        if ($this->getStability() === self::STABLE) {
+            return $versionParser->getMostRecentStable();
+        }
+
+        if ($this->getStability() === self::UNSTABLE) {
+            return $versionParser->getMostRecentUnstable();
+        }
+
+        return $versionParser->getMostRecentAll();
+    }
+
+    /**
+     * Retrieve the current version of the local phar file.
+     *
+     * @param Updater $updater
+     * 
+     * @return string
+     */
+    public function getCurrentLocalVersion(Updater $updater)
+    {
+        return PuliApplicationConfig::VERSION;
+    }
+
+    /**
+     * Set target stability.
+     *
+     * @param string $stability
+     */
+    public function setStability($stability)
+    {
+        if ($stability !== self::STABLE && $stability !== self::UNSTABLE && $stability !== self::ANY) {
```

Comment:
One convenient pattern that I usually follow is to define a private static `$stabilities` and do an `in_array()` check with the strict parameter:

```php
private static $stabilities = array(
    self::STABLE,
    self::UNSTABLE,
    self::ANY,
);

if (!in_array($stability, self::$stabilities, true)) {
    // ...
}
```

This way it's easy to remain all available stabilities in one place.